src/rust/engine/process_execution/src/remote.rs

```diff
@@ -207,121 +208,165 @@ impl CommandRunner {
-              != format!(
-                "type.googleapis.com/{}",
-                precondition_failure.descriptor().full_name()
-              ) {
-              return future::err(ExecutionError::Fatal(format!(
-                "Received FailedPrecondition, but didn't know how to resolve it: {},\
-                 protobuf type {}",
-                execute_response.get_status().get_message(),
-                details.get_type_url()
-              ))).to_boxed() as BoxFuture<_, _>;
-            }
-            try_future!(
-              precondition_failure
-                .merge_from_bytes(details.get_value())
-                .map_err(|e| ExecutionError::Fatal(format!(
-                  "Error deserializing FailedPrecondition proto: {:?}",
-                  e
-                )))
-            );
-
-            let mut missing_digests =
-              Vec::with_capacity(precondition_failure.get_violations().len());
-
-            for violation in precondition_failure.get_violations() {
-              if violation.get_field_type() != "MISSING" {
-                return future::err(ExecutionError::Fatal(format!(
-                  "Didn't know how to process PreconditionFailure violation: {:?}",
-                  violation
-                ))).to_boxed() as BoxFuture<_, _>;
-              }
-              let parts: Vec<_> = violation.get_subject().split("/").collect();
-              if parts.len() != 3 || parts.get(0).unwrap() != &"blobs" {
-                return future::err(ExecutionError::Fatal(format!(
-                  "Received FailedPrecondition MISSING but didn't recognize subject {}",
-                  violation.get_subject()
-                ))).to_boxed() as BoxFuture<_, _>;
-              }
-              let digest = Digest(
-                try_future!(
-                  Fingerprint::from_hex_string(parts.get(1).unwrap()).map_err(|e| {
-                    ExecutionError::Fatal(format!(
-                      "Bad digest in missing blob: {}: {}",
-                      parts.get(1).unwrap(),
-                      e
-                    ))
-                  })
-                ),
-                try_future!(parts.get(2).unwrap().parse::<usize>().map_err(|e| {
-                  ExecutionError::Fatal(format!(
-                    "Missing blob had bad size: {}: {}",
-                    parts.get(2).unwrap(),
-                    e
-                  ))
-                })),
-              );
-              missing_digests.push(digest);
-            }
-            if missing_digests.len() == 0 {
-              return future::err(ExecutionError::Fatal(
-                "Error from remote execution: FailedPrecondition, but no details".to_owned(),
-              )).to_boxed() as BoxFuture<_, _>;;
-            }
-            return future::err(ExecutionError::MissingDigests(missing_digests)).to_boxed()
-              as BoxFuture<_, _>;
+    return stdout;
+  }
+
+  fn extract_stderr(
```

Comment:
We can re-use the logic between `extract_stdout` and `extract_stderr` by taking three functions as arguments here. I think the signature would end up looking like:
`fn extract_output(&self, mut execute_response: ExecuteResponse, Fn(&bazel_protos::remote_execution::ActionResult) -> bool, Fn(&bazel_protos::remote_execution::ActionResult) -> &bazel_protos::remote_execution::Digest, Fn(&bazel_protos::remote_execution::ActionResult) -> bytes::Bytes)`, but there could be some reference lifetime fun. But you'd just call:

```
self.extract_output(
  execute_response,
  |action| action.has_stderr_digest(),
  |action| action.get_stderr_digest()
  |action| action.get_stdout_raw_for_reflect().clone()
)
```

Reply:
I thought about this, but decided against it since readability wise, this didnt look great to me. happy to change though, if you insist.